From 10c578a05d6b6388f3a41571e630de0236f55fc1 Mon Sep 17 00:00:00 2001
From: AdrianDC <radian.dc@gmail.com>
Date: Fri, 1 Jan 2016 15:21:52 +0100
Subject: [PATCH 07/13] Settings: Apps started on boot shortcut in memory
 settings

 * Implement an intent extra to retrieve a tab target
 * Use a tab identifer from apps_ops_categories_cm
    to make sure the selection is correct or ignored

Change-Id: I79d963ecbadc624dc45a398387a348691318b6be
Signed-off-by: AdrianDC <radian.dc@gmail.com>
---
 res/values/cm_strings.xml                               |  3 +++
 res/xml/process_stats_summary.xml                       |  8 +++++++-
 src/com/android/settings/Settings.java                  |  1 +
 .../android/settings/applications/AppOpsSummary.java    | 13 ++++++++++++-
 .../settings/applications/ProcessStatsSummary.java      | 17 +++++++++++++++++
 5 files changed, 40 insertions(+), 2 deletions(-)

diff --git a/res/values/cm_strings.xml b/res/values/cm_strings.xml
index ad0e480..d9ad418 100644
--- a/res/values/cm_strings.xml
+++ b/res/values/cm_strings.xml
@@ -114,6 +114,9 @@
     <!-- Volume link notification -->
     <string name="volume_link_notification_title">Link ring &amp; notification volumes</string>
 
+    <!-- Memory -->
+    <string name="memory_startup_apps_title">Apps started on boot</string>
+
     <!-- Manual provisioning support -->
     <string name="sim_enabler_summary"><xliff:g id="displayName">%1$s</xliff:g> is <xliff:g id="status" example="disabled">%2$s</xliff:g></string>
     <string name="sim_disabled">disabled</string>
diff --git a/res/xml/process_stats_summary.xml b/res/xml/process_stats_summary.xml
index 3b3271d..0e62f5c 100644
--- a/res/xml/process_stats_summary.xml
+++ b/res/xml/process_stats_summary.xml
@@ -1,5 +1,7 @@
 <?xml version="1.0" encoding="utf-8"?>
-<!-- Copyright (C) 2015 The Android Open Source Project
+<!--
+     Copyright (C) 2015 The Android Open Source Project
+     Copyright (C) 2015-2016 The CyanogenMod Project
 
      Licensed under the Apache License, Version 2.0 (the "License");
      you may not use this file except in compliance with the License.
@@ -60,4 +62,8 @@
         android:key="apps_list"
         android:title="@string/memory_usage_apps" />
 
+    <Preference
+        android:key="apps_startup"
+        android:title="@string/memory_startup_apps_title" />
+
 </PreferenceScreen>
diff --git a/src/com/android/settings/Settings.java b/src/com/android/settings/Settings.java
index ee041e8..cbd9995 100644
--- a/src/com/android/settings/Settings.java
+++ b/src/com/android/settings/Settings.java
@@ -1,5 +1,6 @@
 /*
  * Copyright (C) 2008 The Android Open Source Project
+ * Copyright (C) 2016 The CyanogenMod Project
  *
  * Licensed under the Apache License, Version 2.0 (the "License");
  * you may not use this file except in compliance with the License.
diff --git a/src/com/android/settings/applications/AppOpsSummary.java b/src/com/android/settings/applications/AppOpsSummary.java
index 6d817fd..f9528ac 100644
--- a/src/com/android/settings/applications/AppOpsSummary.java
+++ b/src/com/android/settings/applications/AppOpsSummary.java
@@ -1,5 +1,6 @@
 /**
  * Copyright (C) 2013 The Android Open Source Project
+ * Copyright (C) 2016 The CyanogenMod Project
  *
  * Licensed under the Apache License, Version 2.0 (the "License"); you may not
  * use this file except in compliance with the License. You may obtain a copy
@@ -42,6 +43,7 @@ import com.android.internal.logging.nano.MetricsProto.MetricsEvent;
 import com.android.settings.core.InstrumentedPreferenceFragment;
 
 import java.util.ArrayList;
+import java.util.Arrays;
 import java.util.List;
 
 import com.android.settings.development.DevelopmentSettings;
@@ -50,7 +52,7 @@ import com.android.settings.R;
 public class AppOpsSummary extends InstrumentedPreferenceFragment {
     // layout inflater object used to inflate views
     private LayoutInflater mInflater;
-    
+
     private ViewGroup mContentContainer;
     private View mRootView;
     private ViewPager mViewPager;
@@ -133,10 +135,19 @@ public class AppOpsSummary extends InstrumentedPreferenceFragment {
 
         mPageNames = getResources().getTextArray(R.array.app_ops_categories_lineage);
 
+        int defaultTab = -1;
+        Bundle bundle = getArguments();
+        if (bundle != null) {
+            defaultTab = Arrays.asList(mPageNames).indexOf(bundle.getString("appops_tab", ""));
+        }
+
         mViewPager = (ViewPager) rootView.findViewById(R.id.pager);
         mAdapter = new MyPagerAdapter(getChildFragmentManager(),
                 filterTemplates(AppOpsState.ALL_TEMPLATES));
         mViewPager.setAdapter(mAdapter);
+        if (defaultTab >= 0) {
+            mViewPager.setCurrentItem(defaultTab);
+        }
         mViewPager.setOnPageChangeListener(mAdapter);
         PagerTabStrip tabs = (PagerTabStrip) rootView.findViewById(R.id.tabs);
 
diff --git a/src/com/android/settings/applications/ProcessStatsSummary.java b/src/com/android/settings/applications/ProcessStatsSummary.java
index cdb8a4c..8309026 100644
--- a/src/com/android/settings/applications/ProcessStatsSummary.java
+++ b/src/com/android/settings/applications/ProcessStatsSummary.java
@@ -1,5 +1,6 @@
 /*
  * Copyright (C) 2015 The Android Open Source Project
+ * Copyright (C) 2016 The CyanogenMod Project
  *
  * Licensed under the Apache License, Version 2.0 (the "License");
  * you may not use this file except in compliance with the License.
@@ -17,6 +18,7 @@ package com.android.settings.applications;
 
 import android.app.Activity;
 import android.content.Context;
+import android.content.Intent;
 import android.os.Bundle;
 import android.support.v7.preference.Preference;
 import android.support.v7.preference.Preference.OnPreferenceClickListener;
@@ -24,6 +26,7 @@ import android.text.format.Formatter;
 import android.text.format.Formatter.BytesResult;
 import com.android.internal.logging.nano.MetricsProto.MetricsEvent;
 import com.android.settings.R;
+import com.android.settings.Settings.AppOpsSummaryActivity;
 import com.android.settings.SummaryPreference;
 import com.android.settings.Utils;
 import com.android.settings.applications.ProcStatsData.MemInfo;
@@ -38,6 +41,9 @@ public class ProcessStatsSummary extends ProcessStatsBase implements OnPreferenc
     private static final String KEY_AVERAGY_USED = "average_used";
     private static final String KEY_FREE = "free";
     private static final String KEY_APP_LIST = "apps_list";
+    private static final String KEY_APP_STARTUP = "apps_startup";
+
+    private Activity mActivity;
 
     private SummaryPreference mSummaryPref;
 
@@ -46,11 +52,14 @@ public class ProcessStatsSummary extends ProcessStatsBase implements OnPreferenc
     private Preference mAverageUsed;
     private Preference mFree;
     private Preference mAppListPreference;
+    private Preference mAppStartupPreference;
 
     @Override
     public void onCreate(Bundle icicle) {
         super.onCreate(icicle);
 
+        mActivity = getActivity();
+
         addPreferencesFromResource(R.xml.process_stats_summary);
         mSummaryPref = (SummaryPreference) findPreference(KEY_STATUS_HEADER);
         mPerformance = findPreference(KEY_PERFORMANCE);
@@ -59,6 +68,8 @@ public class ProcessStatsSummary extends ProcessStatsBase implements OnPreferenc
         mFree = findPreference(KEY_FREE);
         mAppListPreference = findPreference(KEY_APP_LIST);
         mAppListPreference.setOnPreferenceClickListener(this);
+        mAppStartupPreference = findPreference(KEY_APP_STARTUP);
+        mAppStartupPreference.setOnPreferenceClickListener(this);
     }
 
     @Override
@@ -117,6 +128,12 @@ public class ProcessStatsSummary extends ProcessStatsBase implements OnPreferenc
             startFragment(this, ProcessStatsUi.class.getName(), R.string.memory_usage_apps, 0,
                     args);
             return true;
+        } else if (preference == mAppStartupPreference) {
+            Intent intent = new Intent(Intent.ACTION_MAIN);
+            intent.putExtra("appops_tab", getString(R.string.app_ops_categories_bootup));
+            intent.setClass(mActivity, AppOpsSummaryActivity.class);
+            mActivity.startActivity(intent);
+            return true;
         }
         return false;
     }
-- 
2.7.4

